### PR TITLE
stdenv.mkDerivation: support output checks with and without structuredAttrs

### PR DESCRIPTION
**Pull request by wolfgangwalther**
With a [recent enough nix](https://github.com/NixOS/nix/issues/10856), there will be an eval warning when building a derivation with `__structuredAttrs = true` and e.g. `disallowedReferences = ..`. With structuredAttrs, nix expects those split up in `outputChecks.<output>.disallowedReferences` etc. instead.

To allow derivations to support both structuredAttrs enabled and disabled during a transition period, this PR maps top-level `disallowedReferences` and friends to each output's `outputChecks`. This has the nice side-effect that it is possible to define checks which should apply to *all* outputs at the top-level, and extend them with output-specific checks as well.

While on it, we also apply the same fix that #211783 did to those structured `outputChecks`, which had not been the case before.

Part of the overall effort to enable structuredAttrs by default eventually, tracking: https://github.com/NixOS/nixpkgs/issues/205690

~~Marking as draft for now, because I intend to add some tests as well.~~ It's really annoying to write tests for this, because we effectively need to test a `nix-build` failure :/

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Ma27 — APPROVED** [review]
sgtm :+1:

**Review by philiptaron — APPROVED** [review]
Makes sense to me.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -455,40 +493,16 @@ else let
...
-    } //
-    optionalAttrs (attrs ? allowedRequisites) {
-      allowedRequisites =
-        mapNullable unsafeDerivationToUntrackedOutpath attrs.allowedRequisites;
+    }) // lib.optionalAttrs (!__structuredAttrs) (
+      makeOutputChecks attrs
+    ) // lib.optionalAttrs (__structuredAttrs) {
+      outputChecks = builtins.listToAttrs (map (name: {
+        inherit name;
+        value = lib.zipAttrsWith (_: builtins.concatLists) [
```
Comment (Artturin) [inline pkgs/stdenv/generic/make-derivation.nix:501]: The functions should be inherited from lib at the top of the file for improved perf

**Comment by Artturin** [comment]
Not great performance-wise 
image
(ignore cpuTime)

**Pull request merged**